### PR TITLE
Add "The Road to the JIT"

### PR DESCRIPTION
**Pull request by bjorng**

### CODE REVIEW

File: _posts/2020-12-01-the-road-to-the-jit.md [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+  the next pass.
+
+The later projects mitigated some of the issues in the previous projects.  For example,
+the compilation time was reduced by doing more optimizations before invoking LLVM.
+Ultimately, though, it was decided to terminate the third and final tracing JIT project
+at the end of 2019.
+
+For more information about BEAMJIT, see:
+
+* [BEAMJIT: a just-in-time compiling runtime for Erlang](https://dl.acm.org/doi/abs/10.1145/2633448.2633450?casa_token=tp5GAPlARNcAAAAA%3AEQbE8U2rzCpMKrkzE2eQbbS_fgKuKs7sAuvHFPWO3RjkU1p-3XsMi-Pkbc8A1DFurvvsa6t7qRuI)
```
Comment (sverker) [inline _posts/2020-12-01-the-road-to-the-jit.md:264]: This looks misplaced as the above chapter is about the tracing jits.

Comment (sverker) [reply _posts/2020-12-01-the-road-to-the-jit.md:264]: I take that back. I confused "BEAMJIT" with "BEAMASM".

File: _posts/2020-12-01-the-road-to-the-jit.md [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+Machine). It compiled the Erlang code to C code, which was then
+compiled to native code using GCC.
+
+It was significantly faster than JAM for small projects.
+Unfortunately, compilation was very slow, and the code size of the
+compiled code was too big to make it useful for large projects.
+
+### BEAM (Bogdan's Erlang Abstract Machine)
+
+Bogumil Hausman next machine was called BEAM (Bogdan's Erlang Abstract
```
Comment (ferd) [inline _posts/2020-12-01-the-road-to-the-jit.md]:
```suggestion
Bogumil Hausman's next machine was called BEAM (Bogdan's Erlang Abstract
```

File: _posts/2020-12-01-the-road-to-the-jit.md [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+It was probably at that point we realized that we had a JIT that could
+finally be included in an OTP release.
+
+The next major milestone was reached on May 6 when line numbers in
+stack traces were implemented. That meant that many more test cases now
+succeeded.
+
+Soon after that, all test suites could be run successfully.  During the
+summer and early fall [me](http://github.com/bjorng) and
+[Dan](http://github.com/dgud) joined the project part-time and the
```
Comment (ferd) [inline _posts/2020-12-01-the-road-to-the-jit.md]:
```suggestion
summer and early fall [Dan](http://github.com/dgud) and 
[I](http://github.com/bjorng) joined the project part-time and the
```

File: _posts/2020-12-01-the-road-to-the-jit.md [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+compiler, the OTP team could only afford to spend a limited amount of
+time maintaining HiPE. Therefore, the documentation for HiPE includes
+this note:
+
+> HiPE and execution of HiPE compiled code only have limited support
+  by the OTP team at Ericsson. The OTP team only does limited
+  maintenance of HiPE and does not actively develop HiPE. HiPE is
+  mainly supported by the HiPE team at Uppsala University.
+
+### Other Outcomes the HiPE Project
```
Comment (garazdawi) [inline _posts/2020-12-01-the-road-to-the-jit.md]:
```suggestion
### Other outcomes from the HiPE Project
```

File: _posts/2020-12-01-the-road-to-the-jit.md [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+today if it hasn't been for the HiPE project.  Here are the major
+contributions from the HiPE project to OTP:
+
+* A new [staged tag scheme][newtags] in OTP R7.  The new tag scheme
+  allowed the Erlang system to address the full 4GB address space (the
+  previous tag scheme only supported addressing the lower 1 GB).
+  Surprisingly, the new tag scheme also improved performance.
+
+* The [Core Erlang][core] intermediate representation used in the
+  Erlang compiler to this day.  For more information, see
```
Comment (garazdawi) [inline _posts/2020-12-01-the-road-to-the-jit.md]:
```suggestion
* The [Core Erlang][core] intermediate representation is used in the
  Erlang compiler to this day.  For more information, see
```

File: _posts/2020-12-01-the-road-to-the-jit.md [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+  [An introduction to Core Erlang][core_introduction] and
+  [Core Erlang by Example][core_by_example].
+
+* [Dialyzer][dialyzer] (DIscrepancy AnaLYZer for ERlang programs),
+  started out as a type analysis pass for the HiPE native compiler,
```
Comment (garazdawi) [inline _posts/2020-12-01-the-road-to-the-jit.md:173]:
```suggestion
  started as a type analysis pass for the HiPE native compiler,
```

File: _posts/2020-12-01-the-road-to-the-jit.md [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+  module. The counters were originally meant to be used for finding hot
+  functions and generating native code only for these.  But the overhead
+  in the context switch between interpreted and native code made this
+  usage less useful.
+
+* Repeatedly suggesting that Erlang needed a [literal pool][pool] for premade
+  literal terms (instead of constructing them each time they are
+  used).  At one of our meetings between the HiPE team and the OTP
+  team, I remember [Richard Carlsson][richcarl] pointing out to me that it would
+  be nice for [Wings3D][wings3d] to have floating point literals.  The
```
Comment (garazdawi) [inline _posts/2020-12-01-the-road-to-the-jit.md]:
```suggestion
  be nice for [Wings3D][wings3d] to have floating-point literals.  The
```

File: _posts/2020-12-01-the-road-to-the-jit.md [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+larger than interpreted BEAM code.)
+
+Another change to the design was the tooling for generating the native
+code. In Lukas's prototype, the native code template for each
+instruction was contained in text files similar to the other files
+used by the loader.  That was inflexible, so it was decided to use
+some library that could generate native code.  While some pure C
+libraries could have been used, the C++ library [AsmJIT][asmjit] was
+more convenient in practical use than any of the C libraries.  Also,
+some of C libraries were excluded because they used a GNU license,
```
Comment (garazdawi) [inline _posts/2020-12-01-the-road-to-the-jit.md]:
```suggestion
some C libraries were excluded because they used a GNU license,
```

Comment (garazdawi) [reply _posts/2020-12-01-the-road-to-the-jit.md]: or "some of the"

File: _posts/2020-12-01-the-road-to-the-jit.md [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+
+John joined the practical work on the rejigged JIT project at the end
+of March.
+
+On April 7, 2020, John reached the "prompt beer" milestone.
+
+### Prompt Beer
+
+When the Erlang system is started, a surprisingly large amount of code is
+executed before the prompt appears. On the one hand that means that
```
Comment (garazdawi) [inline _posts/2020-12-01-the-road-to-the-jit.md]:
```suggestion
executed before the prompt appears. On the one hand, that means that
```

**Review by garazdawi — APPROVED** [review]

File: _posts/2020-12-01-the-road-to-the-jit.md (hunk 6 above)
Comment (bjorng) [reply _posts/2020-12-01-the-road-to-the-jit.md:173]: I will keep "started out" as it is closer to my intended meaning than "started". See https://dictionary.cambridge.org/dictionary/english/start-out.

Comment (garazdawi) [reply _posts/2020-12-01-the-road-to-the-jit.md:173]: ok!

**Pull request merged**